Watch face selection using CMake (#1934)
InfiniTimeOrg/InfiniTime: doc/code/Apps.md

## App types

> There are basically 3 types of applications : **system** apps and **user** apps and **watchfaces**.

Reviewer: In documentation we opted to use `watch faces` instead of `watchfaces`. I'd like to keep it consistent. Only if it is a named type, then I'd keep it as one word

```suggestion
There are basically 3 types of applications : **system** apps and **user** apps and **watch faces**.
```

**System** applications are always built into InfiniTime, and InfiniTime cannot work properly without those apps.

> The watch faces, settings, notifications and the application launcher are examples of such system applications.

Reviewer 2: As of this PR, watch faces aren't system applications.

Author: Good catch! I removed watch faces from the list!

**User** applications are optionally built into the firmware. They extend the functionalities of the system.

> **Watchfaces** are very similar to the **user** apps, they are optional, but at least one must be built into the firmware.

Reviewer: ```suggestion
**Watch faces** are very similar to the **user** apps, they are optional, but at least one must be built into the firmware.
```

> The distinction between **system** apps,  **user** apps and watchfaces allows for more flexibility and customization.

Reviewer: two word `watch face` and emphasis (as the other two types are emphasized as well
```suggestion
The distinction between **system** apps,  **user** apps and **watch faces** allows for more flexibility and customization.
```

This allows to easily select which user applications and watchfaces must be built into the firmware
without overflowing the system memory.

> ## Apps and watchfaces initialization

Reviewer: ```suggestion
## Apps and watch faces initialization
```

the corresponding `AppDescription` is first retrieved from `userApps`
and then the function `create` is called to create an instance of the app.

> Watchfaces are handled in a very similar way than the **user** apps : they are created by `DisplayApp` in the method `DisplayApp::LoadScreen()` when the application type is `Apps::Clock`.

Reviewer: space and wording

```suggestion
Watch faces are handled in a very similar way as the **user** apps : they are created by `DisplayApp` in the method `DisplayApp::LoadScreen()` when the application type is `Apps::Clock`.
```

This array `userApps` is used by `DisplayApp` to create the applications and the `AppLauncher`
to list all available applications.

> ## Watchface selection at build time

Reviewer: ```suggestion
## Watch face selection at build time
```

> The list of available watchface is also generated at build time by the `consteval`

Reviewer: ```suggestion
The list of available watch faces is also generated at build time by the `consteval`
```

function `CreateWatchFaceDescriptions()` in `UserApps.h` in the same way as the **user** apps.

> Watchfaces must declare a `WatchFaceTraits` so that the corresponding `WatchFaceDescription` can be generated.

Reviewer: ```suggestion
Watch faces must declare a `WatchFaceTraits` so that the corresponding `WatchFaceDescription` can be generated.
```

$ cmake ... -DENABLE_USERAPPS="Apps::Alarm, Apps::Timer, Apps::MyApp" ... 
```

> Similarly, the list of watchfaces is also generated by CMake, so you need to add the variable `ENABLE_WATCHFACES` to the command line of CMake. It must be set with the list of watchfaces that will be built into the firmware.

Reviewer: `watch face` and one line per sentence, comma separated list
```suggestion
Similarly, the list of watch faces is also generated by CMake, so you need to add the variable `ENABLE_WATCHFACES` to the command line of CMake.
It must be set with the comma separated list of watch faces that will be built into the firmware.
```

> Ex: build the firmware with 3 watchfaces : Analog, PineTimeStyle and Infineat:

Reviewer: ```suggestion
Ex: build the firmware with 3 watch faces : Analog, PineTimeStyle and Infineat:
```
